src/Decomposition.C

```diff
@@ -0,0 +1,206 @@
+    SfcDecomposition::assignKeys(universe, particles);
+    // Sort particles for decomposition
+    std::sort(particles.begin(), particles.end());
+  }
+
+  int findSplitters(BoundingBox &universe, CProxy_Reader &readers) override {
+    BufferedVec<Key> keys;
+
+    // Initial splitter keys (first and last)
+    keys.add(Key(1)); // 0000...1
+    keys.add(~Key(0)); // 1111...1
+    keys.buffer();
+
+    int decomp_particle_sum = 0; // Used to check if all particles are decomposed
+
+    // Main decomposition loop
+    while (keys.size() != 0) {
+      // Send splitters to Readers for histogramming
+      CkReductionMsg *msg;
+      readers.countOct(keys.get(), CkCallbackResumeThread((void*&)msg));
+      int* counts = (int*)msg->getData();
+      int n_counts = msg->getSize() / sizeof(int);
+
+      // Check counts and create splitters if necessary
+      Real threshold = (DECOMP_TOLERANCE * Real(max_particles_per_tp));
+      for (int i = 0; i < n_counts; i++) {
+        Key from = keys.get(2*i);
+        Key to = keys.get(2*i+1);
+
+        int n_particles = counts[i];
+        if ((Real)n_particles > threshold) {
+          // Create 8 more splitter key pairs to go one level deeper.
+          // Leading zeros will be removed in Reader::count() to enable
+          // comparison of splitter key and particle key
+          keys.add(from << 3);
+          keys.add((from << 3) + 1);
+
+          keys.add((from << 3) + 1);
+          keys.add((from << 3) + 2);
+
+          keys.add((from << 3) + 2);
+          keys.add((from << 3) + 3);
+
+          keys.add((from << 3) + 3);
+          keys.add((from << 3) + 4);
+
+          keys.add((from << 3) + 4);
+          keys.add((from << 3) + 5);
+
+          keys.add((from << 3) + 5);
+          keys.add((from << 3) + 6);
+
+          keys.add((from << 3) + 6);
+          keys.add((from << 3) + 7);
+
+          keys.add((from << 3) + 7);
+          if (to == (~Key(0)))
+            keys.add(~Key(0));
+          else
+            keys.add(to << 3);
+        }
+        else {
+          // Create and store splitter
+          Splitter sp(Utility::removeLeadingZeros(from),
+              Utility::removeLeadingZeros(to), from, n_particles);
+          splitters.push_back(sp);
+
+          // Add up number of particles to check if all are flushed
+          decomp_particle_sum += n_particles;
+        }
+      }
+
+      keys.buffer();
+      delete msg;
+    }
+
+    // Check if decomposition is correct
+    if (decomp_particle_sum != universe.n_particles) {
+      CkAbort("Decomposition failure: only %d particles out of %d decomposed",
+          decomp_particle_sum, universe.n_particles);
+    }
+
+    // Sort our splitters
+    std::sort(splitters.begin(), splitters.end());
+
+    // Return the number of TreePieces
+    return splitters.size();
+  }
+};
+
+Decomposition* getDecomposition() {
+    thread_local static std::unique_ptr<Decomposition> decomposition(nullptr);
```

Comment:
There might be a more Charm++ way of having a per-PE variable - maybe use a chare group to hold the decomposition? We can explore this afterwards if you'd like.

Reply:
Alright, I was considering a `group` but wasn't sure if it would be overkill for holding a single pointer; I will change it accordingly.

Reply:
Moved everything to a group, let me know what you think!